mlir/lib/Catalyst/Transforms/DetectQNodes.cpp

```diff
@@ -0,0 +1,1272 @@
+static constexpr llvm::StringRef mlirAsyncRuntimeCreateTokenName = "mlirAsyncRuntimeCreateToken";
+static constexpr llvm::StringRef mlirAsyncRuntimeSetValueErrorName =
+    "mlirAsyncRuntimeSetValueError";
+static constexpr llvm::StringRef mlirAsyncRuntimeSetTokenErrorName =
+    "mlirAsyncRuntimeSetTokenError";
+static constexpr llvm::StringRef mlirAsyncRuntimeIsTokenErrorName = "mlirAsyncRuntimeIsTokenError";
+static constexpr llvm::StringRef mlirAsyncRuntimeIsValueErrorName = "mlirAsyncRuntimeIsValueError";
+static constexpr llvm::StringRef mlirAsyncRuntimeAwaitTokenName = "mlirAsyncRuntimeAwaitToken";
+static constexpr llvm::StringRef mlirAsyncRuntimeAwaitValueName = "mlirAsyncRuntimeAwaitValue";
+static constexpr llvm::StringRef mlirAsyncRuntimeDropRefName = "mlirAsyncRuntimeDropRef";
+
+// Helper function for attributes
+bool hasQnodeAttribute(LLVM::LLVMFuncOp funcOp);
+bool isScheduledForTransformation(LLVM::CallOp callOp);
+bool isAsync(LLVM::LLVMFuncOp funcOp);
+void scheduleCallToInvoke(LLVM::CallOp callOp, PatternRewriter &rewriter);
+void scheduleAnalysisForErrorHandling(LLVM::LLVMFuncOp funcOp, PatternRewriter &rewriter);
+void cleanupPreHandleErrorAttr(LLVM::LLVMFuncOp funcOp, PatternRewriter &rewriter);
+void scheduleLivenessAnalysis(LLVM::CallOp callOp, PatternRewriter &rewriter);
+void annotateCallsForSink(SmallVector<LLVM::CallOp> &calls, PatternRewriter &rewriter);
+bool callsSource(LLVM::CallOp callOp);
+void annotateCallForSource(LLVM::CallOp callOp, PatternRewriter &rewriter);
+void cleanupSink(LLVM::CallOp op, PatternRewriter &rewriter);
+void cleanupSource(LLVM::CallOp source, PatternRewriter &rewriter);
+void cleanupSource(SmallVector<LLVM::CallOp> &sources, PatternRewriter &rewriter);
+void annotateBrToUnreachable(LLVM::BrOp op, PatternRewriter &rewriter);
+
+// Helper function for caller/callees
+LLVM::LLVMFuncOp getCaller(LLVM::CallOp callOp);
+std::optional<LLVM::LLVMFuncOp> getCalleeSafe(LLVM::CallOp callOp);
+
+// Helper functions for matching function names
+bool isFunctionNamed(LLVM::LLVMFuncOp funcOp, llvm::StringRef expectedName);
+bool isAbort(LLVM::LLVMFuncOp funcOp);
+bool callsAbort(LLVM::CallOp callOp);
+bool callsAbort(Operation *possibleCall);
+bool isMlirAsyncRuntimeCreateValue(LLVM::LLVMFuncOp funcOp);
+bool isMlirAsyncRuntimeCreateToken(LLVM::LLVMFuncOp funcOp);
+bool isMlirAsyncRuntimeIsTokenError(LLVM::LLVMFuncOp funcOp);
+bool isMlirAsyncRuntimeIsValueError(LLVM::LLVMFuncOp funcOp);
+bool callsMlirAsyncRuntimeCreateValue(LLVM::CallOp callOp);
+bool callsMlirAsyncRuntimeCreateToken(LLVM::CallOp callOp);
+bool callsMlirAsyncRuntimeCreateToken(Operation *possibleCall);
+bool callsMlirAsyncRuntimeIsTokenError(LLVM::CallOp callOp);
+bool callsMlirAsyncRuntimeIsValueError(LLVM::CallOp callOp);
+bool callsMlirAsyncRuntimeIsTokenError(Operation *possibleCall);
+bool callsMlirAsyncRuntimeIsValueError(Operation *possibleCall);
+bool hasAbortInBlock(Block *block);
```

Comment:
Would you like to consider inlining these helper methods? Would you be able to add them in a header file?

Reply:
I prefer to let the compiler decide what to inline. :)